apps/web/src/lib/api/external/ilmomasiina/actions.ts

```diff
@@ -32,5 +43,143 @@ export async function signUp(formData: FormData) {
   }
 
   revalidateTag("ilmomasiina-events");
-  redirect(`${baseUrl}/signup/${data.id}/${data.editToken}`);
+  redirect(`/${locale}/signups/${data.id}/${data.editToken}`);
+}
+
+const saveSignUpSchema = z
+  .object({
+    signupId: z.string(),
+    signupEditToken: z.string(),
+    firstName: z.string().optional(),
+    lastName: z.string().optional(),
+    namePublic: z.literal("on").optional(),
+    email: z.string().email().optional(),
+  })
+  .and(z.record(z.string().or(z.string().array())));
+
+export async function saveSignUpAction(
+  currentState: unknown,
+  formData: FormData,
+): Promise<{
+  success?: true;
+  errors?: Partial<Record<string, string[]>>;
+} | null> {
+  "use server";
+
+  const locale = getCurrentLocale();
+
+  const formEntries = [...formData.entries()].reduce<
+    Record<string, string | string[]>
+  >((acc, [key, value]) => {
+    if (value instanceof File || value === "" || value.startsWith("$ACTION")) {
+      return acc;
+    }
+
+    if (key in acc) {
+      acc[key] = [value].concat(acc[key]);
+    } else {
+      acc[key] = value;
+    }
+
+    return acc;
+  }, {});
+  const data = saveSignUpSchema.safeParse(formEntries);
+
+  if (!data.success) {
+    return {
+      // TODO: i18n for zod validation errors
+      errors: data.error.flatten().fieldErrors,
+    };
+  }
+
+  const {
+    signupId,
+    signupEditToken,
+    firstName,
+    lastName,
+    email,
+    namePublic,
+    ...otherAnswers
+  } = data.data;
+
+  const t = await getScopedI18n("errors");
+
+  const response = await patchSignUp(signupId, signupEditToken, {
+    id: signupId,
+    answers: Object.entries(otherAnswers).map(([questionId, answer]) => ({
+      questionId,
+      answer,
+    })),
+    language: locale,
+    firstName,
+    lastName,
+    email,
+    namePublic: namePublic === "on",
+  });
+
+  if (!response.ok) {
```

Comment:
Noticed while testing that in the ilmomasiina side while editing a filled out entry, changing name/email is not allowed (expected behavior I think), but when an edit with these fields is sent from the client and next forwards it to ilmomasiina, ilmomasiina responds with a 200 response instead of detecting the invalid changes for the name/email fields. It doesn't update these fields, but doesn't send a 400 with a validation error 😅 

Bug with Ilmomasiina, but we should regardless disable editing the name & email fields on a confirmed signup.

Reply:
Good catch, I recall thinking about this but then forgot

Reply:
Fixed